portfolyo/tools/freq.py

```diff
@@ -48,22 +119,93 @@ def up_or_down(
     1
     >>> freq.up_or_down('MS', 'MS')
     0
+    >>> freq.up_or_down('QS', 'QS-APR')
+    ValueError
+
     """
-    standard_common_ts = pd.Timestamp("2020-01-01 0:00")
-    backup_common_ts = pd.Timestamp("2020-02-03 04:05:06")
-    if common_ts is None:
-        common_ts = standard_common_ts
-
-    ts1 = common_ts + pd.tseries.frequencies.to_offset(freq_source)
-    ts2 = common_ts + pd.tseries.frequencies.to_offset(freq_target)
-    if ts1 > ts2:
-        return 1
-    elif ts1 < ts2:
-        return -1
-    if common_ts == standard_common_ts:
-        # If they are the same, try with another timestamp.
-        return up_or_down(freq_source, freq_target, backup_common_ts)
-    return 0  # only if both give the same answer.
+    # check if passed freuenices are valid
+    # assert_freq_valid(freq_source)
+    # assert_freq_valid(freq_target)
+    # Compare if the freq are the same
+    if freq_source == freq_target:
+        return 0
+    restricted_classes = [
+        pd._libs.tslibs.offsets.QuarterBegin,
+        pd._libs.tslibs.offsets.YearBegin,
+    ]
+    freq_source_as_offset = pd.tseries.frequencies.to_offset(freq_source)
+    freq_target_as_offset = pd.tseries.frequencies.to_offset(freq_target)
+    # One of the freq can be in restricted class, but not both
+    if not (
+        type(freq_source_as_offset) in restricted_classes
+        and type(freq_target_as_offset) in restricted_classes
+    ):
+        try:
+            assert_freq_sufficiently_long(freq_source, freq_target, strict=True)
+            return 1
+        except AssertionError:
+            return -1
+    # If both in restricted class
+    else:
+        source_index = restricted_classes.index(type(freq_source_as_offset))
+        target_index = restricted_classes.index(type(freq_target_as_offset))
+        group_by_month_beginn = (
+            freq_source_as_offset.startingMonth
+            if source_index == 0
+            else freq_source_as_offset.month
+        ) % 3 == (
+            freq_target_as_offset.startingMonth
+            if target_index == 0
+            else freq_target_as_offset.month
+        ) % 3
+
+        if group_by_month_beginn:
+            if source_index > target_index:
+                # we are in case AS and QS
+                return 1
+            elif source_index < target_index:
+                # we are in the case QS and AS
+                return -1
+            elif source_index == 0:
+                # we are in the case QS and QS
+                return 0
+
+        raise ValueError
```

Comment:
Can you add a descriptive message raised with with error, so the user knows what he did wrong?

Reply:
done